### DISCUSSION TITLE
UC1628 parallel driver (128x64)

### DISCUSSION THREAD

**Discussion by ullahsh** [olikraus/u8g2#2011] [General]
I am developing parallel port 8080 driver for UC1628 lcd, do you have any basic driver for parallel port to up the lcd.
wanted to know the state of WR0,WR1,RST and RS pin of LCD to write data into it.
Any help would be appreciated.
Regards,
Shafi

**Comment by olikraus** [comment]
Everything seems to be there in the UC1628 datasheet.
WR0/WR1 --> Page 30 and page 46
RS --> register select, part of the comman tables, called "C/D" on page 13
RST --> should be the usual low level reset input

**Reply by ullahsh** [reply]
Hi olikraus,

Thank you for your quick response. Understood  about the Register select and RST pin
1. Can u please help me with basic parallel port driver if there is any.
2. I have gone through page 30 and 46 but didn't understood it, can u plz tell me the state of WR0 and WR1 pins i should maintain while i write data into LCD.
Regards,
Shafi

**Reply by olikraus** [reply]
From the datasheet attache here: https://github.com/olikraus/u8g2/issues/1066

8080 mode: WR0 = /WR, WR1=/RD
6800 mode: WR0 = R/W, WR1=Enable

In 8080 mode, the WR1 (/RD) signal must be permanently connected to logic high (because reading from display usually is not required). Pull WR0 low to start the write operation.

**Reply by ullahsh** [reply]
I am always setting the RST,RD pin to logic high, and i am changing state CS, RS,WR 
when i write cmd
RS->low, CS-> low, i write cmd on data-pins then WR->high, WR->low, WR->high
when i write data
RS->high,  i write data on data-pins then WR->low, WR->high ,CS->high
 Is this correct?
Should i write LSB first or MSB first? 
![image](https://user-images.githubusercontent.com/118354707/202646595-6373fa73-98ca-43b3-9b0f-3792c030ac05.png)

**Reply by olikraus** [reply]
> Should i write LSB first or MSB first?

Doesn't matter here.

In u8g2 CS is not changed during multi-byte commands. Not sure whether your approch will work.
RS level I don't remember for the UC1628.

**Reply by ullahsh** [reply]
I tried commenting the CS high, but still it's same.

Does **set all pixel on command** forces all pixel on the display to ON?
![image](https://user-images.githubusercontent.com/118354707/203008652-f8e12db2-1880-4da7-8e65-2400349252ce.png)

**Reply by olikraus** [reply]
> Does set all pixel on command forces all pixel on the display to ON?

yes

**Reply by ullahsh** [reply]
Voltage across Vlcd when measure is 0v(pum control is set to Internal Vlcd)...Is it ok?

**Reply by olikraus** [reply]
Good question. Probably not ok.

**Reply by ullahsh** [reply]
ohk got it. I am reading from bias ratio register 0xE8, to do it i am setting WR, RS and resetting RD, read data and setting RD, not able to read it
Is it correct methos to read   
![image](https://user-images.githubusercontent.com/118354707/204460789-91fcd513-77c5-4bce-962e-9b84024ca17d.png)

**Reply by ullahsh** [reply]
Should we expect 6.34 voltage at Vlcd pin before we configure any UC1628c register?

**Reply by olikraus** [reply]
Unknown, I assume this depends a lot on the circuit around the display itself.

**Reply by ullahsh** [reply]
oh....can u please explain the steps to read data from the register example:pump register
![image](https://user-images.githubusercontent.com/118354707/204777652-f1500dae-5874-43b9-bcf3-e23d340c1b83.png)

**Reply by olikraus** [reply]
I never read any data from any display. In fact u8g2 will only write to the display. I don't see the need to read from a display. Anyhow: I can not help with reading the display. It should be described in the data sheet.

**Reply by ullahsh** [reply]
My lcd is up and running, The issue was in hardware for selecting the interface. which was wrongly configured for SPI, instead of parallel 8080. Thanks for the support.

**Comment by bhavyasrig** [comment]
I'm also trying to develop for Parallel port(8080) basic driver. It would be helpful if you can point us to gitlab path if already exists. 
Thanks in advance for the help.

**Reply by olikraus** [reply]
https://github.com/olikraus/u8g2/blob/d23f3543a97bcc95a3f26b5dc8907fb8b2cff227/csrc/u8x8_byte.c#L227-L239

Note: The /WR signal is called "E" in the U8g2 code.

**Comment by fuyuqingfeng** [comment]
Hi olikraus
     I am debugging the 12864 LCD screen based on the uc1628 drive, I want to use the u8g2 library, but I don't see the uc1628 driver file in the source code, which driver file can I use? （uc1601?uc1611?.... ）
     Hope to give me some advice。
     Best Regards

**Reply by fuyuqingfeng** [reply]
I use the 8080 mode for communication

**Reply by olikraus** [reply]
There is no support for the UC1628 at the moment. I also do not know any compatible display controller.

**Reply by olikraus** [reply]
Note: There was an issue for this display: https://github.com/olikraus/u8g2/issues/1066 and https://github.com/olikraus/u8g2/issues/1453
But we were not able to finish this, but this issues includes the datasheet for the display controller.

**Comment by olikraus** [comment]
@fuyuqingfeng @ullahsh @bhavyasrig 
Is there any online marketplace where I could buy a UC1628 LCD?

**Reply by fuyuqingfeng** [reply]
olikraus,You look at this link：
 https://www.topwaydisplay.com/cn/product/mono/LM6020FCW-2

**Reply by olikraus** [reply]
Thanks, looks good but I probably can't order a single display here 🤔

**Reply by bhavyasrig** [reply]
We have ordered via our organization, not sure how to order to one piece seperately.

**Reply by shlyakpavel** [reply]
Hi @olikraus,

Coincidentally, I have a few UC1628 LCD modules in my possession and I'd love to provide you with one of these modules free of charge. I can take care of shipping via DHL for a quick and secure delivery.

Please reach out to me at shlyakpavel@gmail.com with your shipping details, and I'll get everything arranged.

**Reply by olikraus** [reply]
Nice offer. Unfortunately I am currently almost unable to work on u8g2 due to some "real" work which is required here.

**Reply by shlyakpavel** [reply]
@olikraus I can send you one anyway and you can store it until you have more free time or more interest in this particular display. There's no hurry and I would be happy to send you one even if you don't play with it any time soon

**Reply by olikraus** [reply]
Let me participate at the cost. Can you reach me on my e-mail account? See source code.

**Comment by schmidtas** [comment]
Hello everyone,
I would like to take up the topic again.
Were you able to support the library for the UC1628? Or has someone been able to adapt a working file in the meantime?
Thank you in advance for your feedback.

**Comment by olikraus** [comment]
I am still searching for an option to buy once display.

**Reply by schmidtas** [reply]
Dear Oli,
can you send me your Adresse via private Massage.
I'm also located in DE and I will send it directly via DHL.

**Reply by olikraus** [reply]
My email address is there in every .c file and .ino example file...

**Reply by olikraus** [reply]
I have created an issue: https://github.com/olikraus/u8g2/issues/2260

**Comment by olikraus** [comment]
@fuyuqingfeng @ullahsh @bhavyasrig @shlyakpavel 

I have created a 128x64 variant of the UC1628 based display based on the code for a 256x128 display with the same controller.
The implementation may not work yet, but it could be a good starting point. However I need your feedback, whether and how it works as of now. 

A subset of the constructors are also added to all examples:
``` C++
U8G2_UC1628_128X64_1_8080 u8g2(U8G2_R0, 8, 9, 10, 11, 4, 5, 6, 7,  /*enable=*/ 18, /*cs=*/ 3, /*dc=*/ 16, /*reset=*/ 17); // UC1628 pin names: enable: /WR, dc: CD
U8G2_UC1628_128X64_1_4W_SW_SPI u8g2(U8G2_R0, /* clock=*/ 13, /* data=*/ 11, /* cs=*/ 10, /* dc=*/ 9, /* reset=*/ 8);
U8G2_UC1628_128X64_1_4W_HW_SPI u8g2(U8G2_R0, /* cs=*/ 10, /* dc=*/ 9, /* reset=*/ 8);
U8G2_UC1628_256X128_1_4W_SW_SPI u8g2(U8G2_R0, /* clock=*/ 13, /* data=*/ 11, /* cs=*/ 10, /* dc=*/ 9, /* reset=*/ 8);
U8G2_UC1628_256X128_1_4W_HW_SPI u8g2(U8G2_R0, /* cs=*/ 10, /* dc=*/ 9, /* reset=*/ 8);
```
Of course, there is also a 8080 constructor for the 256x128 display, it is just not listed in the examples.

The new code is available in beta release 2.35.6. Maybe someone can test this 128x64 display and give some feedback.
Testing goal: We should be able to see something. Any kind of black pixel (garbage, font, anything) would be great.
You can test both the 256x128 or the 128x64 driver to see something.
As a test, use the ContrastTest and the FlipMode examples from the Arduino IDE.
Please send a picture of at least some black pixel are visible.

Also consider to give feed back in the official issue here: https://github.com/olikraus/u8g2/issues/2260

You can download the latest U8g2 beta release from here: https://github.com/olikraus/U8g2_Arduino/archive/master.zip
Arduino IDE:
1. Remove the existing U8g2_Arduino library (https://stackoverflow.com/questions/16752806/how-do-i-remove-a-library-from-the-arduino-environment)
2. Install the U8g2_Arduino Zip file via Arduino IDE, add zip library menu (https://www.arduino.cc/en/Guide/Libraries).

PlatformIO:
`platformio.ini` (https://docs.platformio.org/en/latest/projectconf/section_env_library.html#lib-deps) should include 
```
lib_deps =
  u8g2=https://github.com/olikraus/U8g2_Arduino/archive/master.zip
```
